picard/tagger.py

```diff
@@ -512,80 +544,122 @@ def add_files(self, filenames, target=None):
+                self._pending_files_process_queue.put(file.filename)
+
+    def _add_files_thread(self):
+        import time
+        while True:
+            filename, metadata = self._pending_files_results_queue.get()
+
+            if self.stopping:
+                break
+
+            if filename in self._pending_files:
+                file, target = self._pending_files.pop(filename)
+                if target is None or target is self.unclustered_files:
+                    thread.to_main(self.unclustered_files.add_files, [file])
+                    time.sleep(0.005)
+                self._pending_load_finished_queue.put((file, metadata, target))
+                continue
 
-    def add_directory(self, path):
-        if config.setting['recursively_add_files']:
-            self._add_directory_recursive(path)
-        else:
-            self._add_directory_non_recursive(path)
+        # Append empty message to kill _file_load_finished_thread and _file_loaded_thread
+        self._pending_load_finished_queue.put((None, None, None))
 
-    def _add_directory_recursive(self, path):
-        ignore_hidden = config.setting["ignore_hidden_files"]
-        walk = os.walk(path)
+    def _file_load_finished_thread(self):
+        import time
+        while True:
+            file, metadata, target = self._pending_load_finished_queue.get()
 
-        def get_files():
-            try:
-                root, dirs, files = next(walk)
-                if ignore_hidden:
-                    dirs[:] = [d for d in dirs if not is_hidden(os.path.join(root, d))]
-            except StopIteration:
-                return None
+            if self.stopping:
+                break
+
+            if file:
+                while not self.stopping and file not in self.unclustered_files.files:
+                    time.sleep(0)
+                if not metadata:
+                    error = True
+                else:
+                    error = None
+                thread.to_main(file._loading_finished, result=metadata, error=error, callback=None)
+
+                # Delay next loading finished for longer periods
+                #   while other tasks are still running
+                if not self._pending_files_results_queue.empty():
+                    time.sleep(0.5)
+                else:
+                    time.sleep(0.01)
+                self._file_loaded_queue.put((file, target))
+
+        # Append empty message to kill _file_loaded_thread
+        self._file_loaded_queue.put((None, None))
+
+    def _file_loaded_thread(self):
+        import time
+        while True:
+            file, target = self._file_loaded_queue.get()
+
+            if self.stopping:
+                break
+
+            thread.to_main(self._file_loaded, file, target)
+            if not self._pending_files_results_queue.empty():
+                time.sleep(5)
+            elif not self._file_loaded_queue.empty():
+                time.sleep(2)
```

Comment:
I don't get the exact purpose of all those `sleep()` and/or how values were chosen.

Reply:
To prioritize threads jobs. Just to make sure they yield their time slice and let the UI and other threads do their job first.
Values were experimental, tuned manually to prevent stuttering while maintaining a sense of progress for all the different tasks.

Reply:
The purpose of Sleep is to ensure that the GIL is freed up for long enough without the same thread grabbing it back again to ensure that a different thread gets the GIL. However it does NOT ensure that the main thread gets the GIL and it is entirely possible for e.g. two worker threads which do micro-sleeps to swap the GIL back and forth without the main thread getting a look in. Additionally, when there is no other work you are sleeping for no real reason.

My assessment is that the best means - by which I mean it is deterministic in allowing the main thread to get a look in and it doesn't waste time in unnecessary micro-sleeps - is for worker threads to use a Python Signal to create an event on the main thread, and then to wait on a Queue. It is possible that another worker thread gets the GIL, but eventually that will also signal and wait on the Queue. Eventually all worker threads will be in a wait state and the main thread will definitely execute and process the Signals. For each signal it processes, the main thread should put a dummy entry on the queue, and this will then unblock the worker thread to execute again. This is deterministic, and handled completely automatically by Python.

Reply:
And you're absolutely correct on that, but that's way too much trouble and you won't get any more performance from that, more form than function. 

Speeding up the startup is my next quest. It takes up to a minute to show anything on the screen, which is insane. Only happens on my desktop, which probably means it isn't directly related to Picard.

Reply:
What you get is stability.

Reply:
Discovered that the startup was slowed down by stupid OneDrive trying to sync the desktop.ini file.
![image](https://user-images.githubusercontent.com/1678454/81479738-8a388480-91fb-11ea-9c5d-9bffcd5c5a3c.png)